### PR TITLE
Use `BufferStream` as a sacrificial I/O

### LINKED ISSUE #474: Truncated data from S3 when using AWS v1.63
Here's a public dataset for which `AWSS3.s3_get_file` returns almost no data when using AWS.jl v1.63 but works as  expected with v1.62 
```sh
$ aws s3 ls "s3://invenia-public-datasets/Impute/v1.0.0/datasets.tar.gz" 
2020-10-10 20:52:15    5082788 datasets.tar.gz
```
### Using v1.63.0
```jl
(jl_u1dtGQ) pkg> st
      Status `/private/var/folders/hx/1h0bbkfd18d4n1qrnwmrl4j00000gn/T/jl_u1dtGQ/Project.toml`
  [fbe9abb3] AWS v1.63.0
  [1c724243] AWSS3 v0.9.2
  
julia> using AWS, AWSS3

julia> s3p = S3Path("s3://invenia-public-datasets/Impute/v1.0.0/datasets.tar.gz");

julia> s3_get_file(s3p.bucket, s3p.key, "163.tar.gz")

julia> stat("163.tar.gz")  # just 3 bytes
StatStruct(mode=0o100644, size=3)   
```
### Using v1.62.2
```jl
(jl_0F6HFC) pkg> st
      Status `/private/var/folders/hx/1h0bbkfd18d4n1qrnwmrl4j00000gn/T/jl_0F6HFC/Project.toml`
  [fbe9abb3] AWS v1.62.2
  [1c724243] AWSS3 v0.9.2

julia> s3p = S3Path("s3://invenia-public-datasets/Impute/v1.0.0/datasets.tar.gz");

julia> s3_get_file(s3p.bucket, s3p.key, "162.tar.gz")

julia> stat("162.tar.gz")  # expected size
StatStruct(mode=0o100644, size=5082788)
```

### PR DESCRIPTION
**Pull request by omus**
Fixes #474

### CODE REVIEW

**Comment by omus** [comment]
bors try

File: src/utilities/request.jl [hunk 1]
```diff
@@ -172,25 +172,13 @@ function _http_request(http_backend::HTTPBackend, request::Request, response_str
...
         finally
-            # Read directly from the buffer's data as the I/O stream is closed by HTTP.jl.
-            # We need to be careful to always write the buffer data to the `response_stream`
-            # when an exception occurs as this data contains details about the AWS error.
-            data = if isopen(buffer)
-                take!(buffer)
-            else
-                # When closed the `IOBuffer` size will be set to 0. In order to retrieve the
-                # raw data we need to determine the size from the first '\0' character.
-                first_null = findfirst(==(0x00), buffer.data)
```
Comment (omus) [inline src/utilities/request.jl:183]: This was a naive oversight on my part

File: test/issues.jl [hunk 2]
```diff
@@ -124,6 +124,37 @@ try
         end
     end
 
+    @testset "issue 474" begin
+        body = "foo\0bar"
```
Comment (mattBrzezinski) [inline test/issues.jl:128]:
```suggestion
        body = "foo\0bar"
        expected = Vector{UInt8}(body)
```

File: test/issues.jl [hunk 3]
```diff
@@ -124,6 +124,37 @@ try
...
+            @test raw isa Vector{UInt8}
+            @test raw == b"foo\0bar"
+
+            stream = S3.get_object(BUCKET_NAME, file_name, Dict("return_stream" => true))
+            if AWS.DEFAULT_BACKEND[] isa AWS.HTTPBackend
+                @test stream isa Base.BufferStream
+                @test !isopen(stream)
+
+                if !isopen(stream)
+                    @test read(stream) == b"foo\0bar"
```
Comment (mattBrzezinski) [inline test/issues.jl:144]:
```suggestion
                    @test read(stream) == expected
```

File: test/issues.jl [hunk 4]
```diff
@@ -124,6 +124,37 @@ try
...
+                end
+            else
+                @test stream isa IOBuffer
+                @test isopen(stream)
+
+                seekstart(stream)
+                @test read(stream) == b"foo\0bar"
```
Comment (mattBrzezinski) [inline test/issues.jl:151]:
```suggestion
                @test read(stream) == expected
```

File: test/minio.jl [hunk 5]
```diff
@@ -104,6 +105,37 @@ try
         end
     end
 
+    @testset "issue 474" begin
+        body = "foo\0bar"
```
Comment (mattBrzezinski) [inline test/minio.jl:109]:
```suggestion
        body = "foo\0bar"
        expected = Vector{UInt8}(body)
```

File: test/minio.jl [hunk 6]
```diff
@@ -104,6 +105,37 @@ try
...
+        file_name = "null.txt"
+        bucket_name = "anewbucket"
+
+        try
+            S3.put_object(bucket_name, file_name, Dict("body" => body))
+
+            raw = S3.get_object(bucket_name, file_name, Dict("return_raw" => true))
+            @test raw isa Vector{UInt8}
+            @test raw == b"foo\0bar"
```
Comment (mattBrzezinski) [inline test/minio.jl:118]:
```suggestion
            @test raw == expected
```

File: test/minio.jl [hunk 7]
```diff
@@ -104,6 +105,37 @@ try
...
+
+            stream = S3.get_object(bucket_name, file_name, Dict("return_stream" => true))
+            if AWS.DEFAULT_BACKEND[] isa AWS.HTTPBackend
+                @test stream isa Base.BufferStream
+                @test !isopen(stream)
+
+                if !isopen(stream)
+                    @test read(stream) == b"foo\0bar"
```
Comment (mattBrzezinski) [inline test/minio.jl:126]:
```suggestion
                    @test read(stream) == expected
```

File: test/minio.jl [hunk 8]
```diff
@@ -104,6 +105,37 @@ try
...
+                end
+            else
+                @test stream isa IOBuffer
+                @test isopen(stream)
+                seekstart(stream)
+                @test read(stream) == b"foo\0bar"
```
Comment (mattBrzezinski) [inline test/minio.jl:132]:
```suggestion
                @test read(stream) == expected
```

File: src/utilities/request.jl [hunk 9]
```diff
@@ -172,25 +172,13 @@ function _http_request(http_backend::HTTPBackend, request::Request, response_str
...
-                buffer.data[firstindex(buffer.data):last_index]
-            end
+            # We're unable to read from the `Base.BufferStream` until it has been closed.
+            # HTTP.jl will close passed in `response_stream` keyword. This ensures that it
+            # is always closed (e.g. HTTP.jl 0.9.15)
+            close(buffer)
 
-            write(response_stream, data)
+            # Transfer the contents of the `BufferStream` into `response_stream` variable.
+            write(response_stream, read(buffer))
```
Comment (oxinabox) [inline src/utilities/request.jl:181]: I have a feeling there is a more efficient way of doing this,
that avoids the intermediate vector that `read` will allocate

File: test/minio.jl [hunk 10]
```diff
@@ -65,6 +65,7 @@ try
     # Duplicated testset from "test/issues.jl". Useful for testing outside the CI. Ideally,
```
Comment (mattBrzezinski) [inline test/minio.jl:65]: Out-of-scope, we should remove this duplication and centralize test cases between MinIO and non.

**Review by mattBrzezinski — APPROVED** [review]
LGTM, very minor proposal to remove some duplication.

File: src/utilities/request.jl (hunk 9 above)
Comment (oxinabox) [reply src/utilities/request.jl:181]: I think it is
```suggestion
            write(response_stream, buffer)
```

Comment (omus) [reply src/utilities/request.jl:181]: Confirmed this works with the tests I can run locally

**Comment by omus** [comment]
bors try

**Pull request merged**